### ISSUE TITLE
Authentication Question for Single Sign On

### ISSUE THREAD

**Issue by MaxHampton** [OneDrive/onedrive-sdk-csharp#9]
I am currently developing a cross-platform app using Xamarin, that targets Windows Phone 8.1, Android, and iOS. I am using a PCL to handle all of the OneDrive specific code, and have an object in each platform project that acts as an adapter for the OneDriveClient.

In my WP project, I am able to sign in using this client:

```
_client = OneDriveClient.GetMicrosoftAccountClient(
    appId: Resources.MicrosoftOneDriveClientId,
    returnUrl: Resources.MicrosoftOneDriveReturnUrl,
    scopes: new[] {"wl.signin", "wl.offline_access", "onedrive.readwrite"},
    clientSecret: Resources.MicrosoftOneDriveClientSecret,
    credentialCache: credentialCache,
    httpProvider: null,
    serviceInfoProvider: new ServiceInfoProvider(webAuthenticationUi));
```

I am instantiating credentialCache as follows:

```
CredentialCache credentialCache;
if (!_resourceService.TryLoadOneDriveCredentialCache(out credentialCache))
{
    credentialCache = new CredentialCache();
}
```

The TryLoad method reads the byte[] from CredentialCache::GetCacheBlob() that I have written to app resources, after the CredentialCache::AfterAccess delegate fires. In app resources it would look like 

```
{ OneDriveCredentialCacheKey, byte[] }
```

This all works great, and after calling _client.AuthenticateAsync() I am able to query my OneDrive drive. However, on subsequent sessions of the app, when I call AuthenticateAsync() with the loaded credentialCache, I am still being presented with the webAuthenticationUi.

I assume this comes down to how I'm instantiating the OneDriveClient or the CredentialCache. Is there a different signature I should use when I have a valid CredentialCache vs. not? Do I need to implement any of the other Interfaces and use them instead of the defaults?

My goal is to enable single sign on across the different platforms. Thanks for any feedback / pointers in the right direction.

**Comment by ginach** [comment]
Hi Max,

The cache contains all of the stored account sessions but the authentication provider also needs to know which account to load during the first auth of the session to pull from the cache. The way you can specify this currently is by setting _client.ServiceInfo.UserId to the user ID you'd like to load from the cache. Authentication will invoke the IWebAuthenticationUi if this value isn't found, resulting in the behavior you mention.

This was a miss with the current design and I'm working on a way to specify this during client creation.

**Comment by ginach** [comment]
Hi Max,

The cache contains all of the stored account sessions but the authentication provider also needs to know which account to load during the first auth of the session to pull from the cache. It looks like there's currently a bug not allowing the user ID to be specified at time of authentication. A workaround for now is to implement your own IServiceInfoProvider that inherits from ServiceInfoProvider that sets the UserId in GetServiceInfo after calling base:

public override Task<ServiceInfo> GetServiceInfo(
            AppConfig appConfig,
            CredentialCache credentialCache,
            IHttpProvider httpProvider,
            ClientType clientType)
        {
            var serviceInfo = await base.GetServiceInfo(appConfig, credentialCache, httpProvider, clientType);
            serviceInfo.UserId = {user ID};
            return serviceInfo;
        }

I'll work on a fix to specify this during client creation.

**Comment by MaxHampton** [comment]
Workaround seems to be working, thanks very much!

**Comment by mybluedog** [comment]
This is working for me too.  Right now, I am parsing the credentials blob to retrieve the User ID (since the User ID is hidden in a private dictionary).  Is this the best way to get the ID currently?  I assume that the fix will make this unnecssary somehow?

**Comment by ginach** [comment]
Correct, when I fix this there will be a way to specify user ID without creating your own ServiceInfoProvider.

**Comment by ginach** [comment]
There's a fix in version 1.1.29 that allows setting this via a property on the ServiceInfoProvider now. It's not a full fix, but should make it so you don't have to create a whole new class. An example calling pattern would be:

OneDriveClient.GetMicrosoftAccountClient(appId, returnUrl, scopes, serviceInfoProvider: new ServiceInfoProvider { UserSignInName = "emailAddress" };

**Comment by baywet** [comment]
Thank you for reaching out and for your patience. This SDK is being officially deprecated. See #259 for more information
